cpp/dolfin/fem/DofMapPermuter.cpp

```diff
@@ -0,0 +1,459 @@
+  throw std::runtime_error("Two of a triangle's vertices appear to be equal.");
+}
+//-----------------------------------------------------------------------------
+/// Calculates the permutation orders for a tetrahedron
+/// @param[in] v1 The global vertex number of the tetrahedron's first vertex
+/// @param[in] v2 The global vertex number of the tetrahedron's second vertex
+/// @param[in] v3 The global vertex number of the tetrahedron's third vertex
+/// @param[in] v4 The global vertex number of the tetrahedron's fourth vertex
+/// @return The rotation and reflection orders for the tetrahedron
+std::array<int, 4> _calculate_tetrahedron_orders(int v1, int v2, int v3, int v4)
+{
+  if (v1 < v2 && v1 < v3 && v1 < v4)
+  {
+    auto tri_orders = _calculate_triangle_orders(v2, v3, v4);
+    return {0, 0, tri_orders[0], tri_orders[1]};
+  }
+  if (v2 < v1 && v2 < v3 && v2 < v4)
+  {
+    auto tri_orders = _calculate_triangle_orders(v3, v1, v4);
+    return {1, 0, tri_orders[0], tri_orders[1]};
+  }
+  if (v3 < v1 && v3 < v2 && v3 < v4)
+  {
+    auto tri_orders = _calculate_triangle_orders(v1, v2, v4);
+    return {2, 0, tri_orders[0], tri_orders[1]};
+  }
+  if (v4 < v1 && v4 < v2 && v4 < v3)
+  {
+    auto tri_orders = _calculate_triangle_orders(v2, v1, v3);
+    return {0, 1, tri_orders[0], tri_orders[1]};
+  }
+  throw std::runtime_error(
+      "Two of a tetrahedron's vertices appear to be equal.");
+}
+//-----------------------------------------------------------------------------
+/// Makes a permutation to flip the dofs on an edge
+/// @param[in] edge_dofs Number of edge dofs
+/// @param[in] blocksize Block size
+/// @return The permutation to reverse the dofs
+std::vector<int> _edge_flip(const int edge_dofs, const int blocksize)
+{
+  // This will only be called at most once for each mesh
+  const int blocks = edge_dofs / blocksize;
+  int j = 0;
+  std::vector<int> flip(edge_dofs);
+  for (int i = 0; i < blocks; ++i)
+    for (int k = 0; k < blocksize; ++k)
+      flip[j++] = edge_dofs - (i + 1) * blocksize + k;
+  return flip;
+}
+//-----------------------------------------------------------------------------
+/// Makes permutations to rotate and reflect the dofs on a triangle
+/// @param[in] edge_dofs Number of dofs on the face of the triangle
+/// @return Permutations to rotate and reflect the dofs
+std::array<std::vector<int>, 2>
+_triangle_rotation_and_reflection(const int face_dofs, const int blocksize)
+{
+  // This will only be called at most once for each mesh
+  const int blocks = face_dofs / blocksize;
+  float root = std::sqrt(8 * blocks + 1);
+  assert(root == std::floor(root) && int(root) % 2 == 1);
+  int side_length = (root - 1) / 2; // side length of the triangle of face dofs
+
+  std::vector<int> rotation(face_dofs);
+  {
+    int j = 0;
+    int i = 1;
+    for (int st = blocks - 1; st >= 0; st -= (i++))
+    {
+      int dof = st;
+      for (int sub = i + 1; sub <= side_length + 1; dof -= (sub++))
```

Comment:
Similarly, maybe do the operation on `dof` outside the `for` statement.

Reply:
I've changed all three loops like this